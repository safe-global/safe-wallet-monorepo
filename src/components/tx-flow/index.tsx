import { createContext, type ReactElement, type ReactNode, useState, useEffect, useCallback } from 'react'
import TxModalDialog from '@/components/common/TxModalDialog'
import { usePathname } from 'next/navigation'
import useSafeInfo from '@/hooks/useSafeInfo'

const noop = () => {}

type TxModalContextType = {
  txFlow: JSX.Element | undefined
  setTxFlow: (txFlow: TxModalContextType['txFlow'], onClose?: () => void, shouldWarn?: boolean) => void
  setFullWidth: (fullWidth: boolean) => void
}

export const TxModalContext = createContext<TxModalContextType>({
  txFlow: undefined,
  setTxFlow: noop,
  setFullWidth: noop,
})

export const TxModalProvider = ({ children }: { children: ReactNode }): ReactElement => {
  const [txFlow, setFlow] = useState<TxModalContextType['txFlow']>(undefined)
  const [shouldWarn, setShouldWarn] = useState<boolean>(true)
  const [, setOnClose] = useState<Parameters<TxModalContextType['setTxFlow']>[1]>(noop)
  const [fullWidth, setFullWidth] = useState<boolean>(false)
  const pathname = usePathname()
  const [, setLastPath] = useState<string>(pathname)
  const { safeAddress, safe } = useSafeInfo()

  const handleModalClose = useCallback(() => {
    setOnClose((prevOnClose) => {
      prevOnClose?.()
      return noop
    })
    setFlow(undefined)
  }, [setFlow, setOnClose])

  const handleShowWarning = useCallback(() => {
    if (!shouldWarn) {
      handleModalClose()
      return
    }

    const ok = confirm('Closing this window will discard your current progress.')
    if (ok) {
      handleModalClose()
    }
  }, [shouldWarn, handleModalClose])

  const setTxFlow = useCallback(
    (txFlow: TxModalContextType['txFlow'], onClose?: () => void, shouldWarn?: boolean) => {
      setFlow(txFlow)
      setOnClose(() => onClose ?? noop)
      setShouldWarn(shouldWarn ?? true)
    },
    [setFlow, setOnClose],
  )

  // Show the confirmation dialog if user navigates
  useEffect(() => {
    setLastPath((prev) => {
      if (prev !== pathname && txFlow) {
        handleShowWarning()
      }
      return pathname
    })
  }, [txFlow, handleShowWarning, pathname])

  // Close the modal when the Safe changes
  useEffect(() => {
    handleModalClose()
    // Could have same address but different chain
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [safe.chainId, safeAddress])

  return (
    <TxModalContext.Provider value={{ txFlow, setTxFlow, setFullWidth }}>
      {children}

      <TxModalDialog open={!!txFlow} onClose={handleShowWarning} fullWidth={fullWidth}>
        {txFlow}
      </TxModalDialog>
    </TxModalContext.Provider>
  )
}
