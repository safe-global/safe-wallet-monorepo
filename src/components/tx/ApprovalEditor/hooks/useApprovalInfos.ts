import useAsync from '@/hooks/useAsync'
import useBalances from '@/hooks/useBalances'
import { ApprovalModule, type ApprovalModuleResponse } from '@/services/security/modules/ApprovalModule'
import type { SecurityResponse } from '@/services/security/modules/types'
import { getERC20TokenInfoOnChain, UNLIMITED_APPROVAL_AMOUNT } from '@/utils/tokens'
import { type SafeTransaction } from '@safe-global/safe-core-sdk-types'
import { type TokenInfo } from '@safe-global/safe-gateway-typescript-sdk'
import { ethers } from 'ethers'
import { PSEUDO_APPROVAL_VALUES } from '../utils/approvals'

export type ApprovalInfo = {
  tokenInfo: (Omit<TokenInfo, 'logoUri' | 'name'> & { logoUri?: string }) | undefined
  tokenAddress: string
  spender: any
  amount: any
  amountFormatted: string
}

const ApprovalModuleInstance = new ApprovalModule()

const useApprovalData = (safeTransaction: SafeTransaction | undefined) => {
  return useAsync<SecurityResponse<ApprovalModuleResponse>>(() => {
    if (!safeTransaction) {
      return
    }

    return ApprovalModuleInstance.scanTransaction({ safeTransaction })
  }, [safeTransaction])
}

export const useApprovalInfos = (safeTransaction: SafeTransaction | undefined) => {
  const [approvals] = useApprovalData(safeTransaction)

  const { balances } = useBalances()

  return useAsync<ApprovalInfo[]>(
    async () => {
      if (!approvals || !approvals.payload || approvals.payload.length === 0) return Promise.resolve([])

      return Promise.all(
        approvals.payload.map(async (approval) => {
          let tokenInfo: Omit<TokenInfo, 'name' | 'logoUri'> | undefined = balances.items.find(
            (item) => item.tokenInfo.address === approval.tokenAddress,
          )?.tokenInfo
          if (!tokenInfo) {
            tokenInfo = await getERC20TokenInfoOnChain(approval.tokenAddress)
          }

          const amountFormatted = UNLIMITED_APPROVAL_AMOUNT.eq(approval.amount)
            ? PSEUDO_APPROVAL_VALUES.UNLIMITED
            : ethers.utils.formatUnits(approval.amount, tokenInfo?.decimals)

          return { ...approval, tokenInfo: tokenInfo, amountFormatted }
        }),
      )
    },
    [balances.items.length, approvals],
    false, // Do not clear data on balance updates
  )
}
