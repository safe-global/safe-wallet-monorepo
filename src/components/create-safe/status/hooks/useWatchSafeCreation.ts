import { PendingSafeData } from '@/components/create-safe'
import { Dispatch, SetStateAction, useEffect } from 'react'
import { useRouter } from 'next/router'
import { pollSafeInfo } from '@/components/create-safe/status/usePendingSafeCreation'
import { AppRoutes } from '@/config/routes'
import { SafeCreationStatus } from '@/components/create-safe/status/useSafeCreation'
import { trackEvent, CREATE_SAFE_EVENTS, SAFE_APPS_EVENTS } from '@/services/analytics'
import { useAppSelector } from '@/store'
import { selectChainById } from '@/store/chainsSlice'

const useWatchSafeCreation = ({
  status,
  safeAddress,
  pendingSafe,
  setPendingSafe,
  setStatus,
  chainId,
}: {
  status: SafeCreationStatus
  safeAddress: string | undefined
  pendingSafe: PendingSafeData | undefined
  setPendingSafe: Dispatch<SetStateAction<PendingSafeData | undefined>>
  setStatus: Dispatch<SetStateAction<SafeCreationStatus>>
  chainId: string
}) => {
  const router = useRouter()
  const chain = useAppSelector((state) => selectChainById(state, chainId))

  useEffect(() => {
    const checkCreatedSafe = async (chainId: string, address: string) => {
      try {
        await pollSafeInfo(chainId, address)
        setStatus(SafeCreationStatus.INDEXED)
      } catch (e) {
        setStatus(SafeCreationStatus.INDEX_FAILED)
      }
    }

    if (status === SafeCreationStatus.INDEXED) {
      trackEvent(CREATE_SAFE_EVENTS.GET_STARTED)
      const chainPrefix = chain?.shortName

      if (safeAddress && chainPrefix) {
        const address = `${chainPrefix}:${safeAddress}`
        const redirectUrl = router.query?.safeViewRedirectURL
        if (typeof redirectUrl === 'string') {
          // We're prepending the safe address directly here because the `router.push` doesn't parse
          // The URL for already existing query params
          const hasQueryParams = redirectUrl.includes('?')
          const appendChar = hasQueryParams ? '&' : '?'

          if (redirectUrl.includes('apps')) {
            trackEvent({ ...SAFE_APPS_EVENTS.SHARED_APP_OPEN_AFTER_SAFE_CREATION })
          }

          router.push(redirectUrl + `${appendChar}safe=${address}`)
        } else {
          router.push({ pathname: AppRoutes.safe.home, query: { safe: address } })
        }
      }
    }

    if (status === SafeCreationStatus.SUCCESS) {
      trackEvent(CREATE_SAFE_EVENTS.CREATED_SAFE)

      safeAddress && pendingSafe && checkCreatedSafe(chainId, safeAddress)
      setPendingSafe(undefined)
    }

    if (status === SafeCreationStatus.ERROR || status === SafeCreationStatus.REVERTED) {
      if (pendingSafe?.txHash) {
        setPendingSafe((prev) => (prev ? { ...prev, txHash: undefined } : undefined))
      }
    }
  }, [router, safeAddress, setPendingSafe, status, pendingSafe, setStatus, chainId, chain])
}

export default useWatchSafeCreation
