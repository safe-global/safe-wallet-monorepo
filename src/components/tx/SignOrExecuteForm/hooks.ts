import { useMemo } from 'react'
import { type TransactionOptions, type SafeTransaction } from '@safe-global/safe-core-sdk-types'
import useSafeInfo from '@/hooks/useSafeInfo'
import useWallet from '@/hooks/wallets/useWallet'
import useOnboard from '@/hooks/wallets/useOnboard'
import { isSmartContractWallet } from '@/hooks/wallets/wallets'
import {
  dispatchOnChainSigning,
  dispatchTxExecution,
  dispatchTxProposal,
  dispatchTxSigning,
} from '@/services/tx/tx-sender'
import { useHasPendingTxs } from '@/hooks/usePendingTxs'
import { sameString } from '@safe-global/safe-core-sdk/dist/src/utils'

export const useTxActions = () => {
  const { safe, safeAddress } = useSafeInfo()
  const { chainId, version } = safe
  const onboard = useOnboard()
  const wallet = useWallet()

  return useMemo(() => {
    const proposeTx = async (sender: string, safeTx: SafeTransaction, txId?: string, origin?: string) => {
      const tx = await dispatchTxProposal({
        chainId,
        safeAddress,
        sender,
        safeTx,
        txId,
        origin,
      })
      return tx.txId
    }

    return {
      async signTx(safeTx: SafeTransaction, txId?: string): Promise<string> {
        if (!onboard) throw 'Onboard not ready'
        if (!wallet) throw 'Wallet not connected'

        // Smart contract wallets must sign via an on-chain tx
        if (await isSmartContractWallet(wallet)) {
          // If the first signature is a smart contract wallet, we have to propose w/o signatures
          // Otherwise the backend won't pick up the tx
          // The signature will be added once the on-chain signature is indexed
          const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin))
          await dispatchOnChainSigning(safeTx, id, onboard, chainId)
          return id
        }

        // Otherwise, sign off-chain
        const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId)
        return await proposeTx(wallet.address, signedTx, txId, origin)
      },

      async executeTx(
        safeTx: SafeTransaction,
        txId: string | undefined,
        txOptions: TransactionOptions,
      ): Promise<string> {
        if (!onboard) throw 'Onboard not ready'
        if (!wallet) throw 'Wallet not connected'

        const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin))
        await dispatchTxExecution(safeTx, txOptions, id, onboard, chainId)
        return id
      },
    }
  }, [chainId, safeAddress, version, onboard, wallet])
}

export const useValidateNonce = (safeTx?: SafeTransaction): boolean => {
  const { safe } = useSafeInfo()
  return !!safe && !!safeTx && safeTx?.data.nonce === safe.nonce
}

export const useImmediatelyExecutable = (): boolean => {
  const { safe } = useSafeInfo()
  const hasPending = useHasPendingTxs()
  return safe && safe.threshold === 1 && !hasPending
}

// Check if the executor is the safe itself (it won't work)
export const useIsExecutionLoop = (): boolean => {
  const wallet = useWallet()
  const { safeAddress } = useSafeInfo()
  return wallet ? sameString(wallet.address, safeAddress) : false
}
