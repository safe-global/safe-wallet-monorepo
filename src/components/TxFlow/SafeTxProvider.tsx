import { createContext, useState, useEffect } from 'react'
import type { Dispatch, ReactNode, SetStateAction, ReactElement } from 'react'
import type { SafeTransaction } from '@safe-global/safe-core-sdk-types'
import { createTx } from '@/services/tx/tx-sender'
import { useRecommendedNonce } from '../tx/SignOrExecuteForm/hooks'

export const SafeTxContext = createContext<{
  safeTx?: SafeTransaction
  setSafeTx: Dispatch<SetStateAction<SafeTransaction | undefined>>

  safeTxError?: Error
  setSafeTxError: Dispatch<SetStateAction<Error | undefined>>

  nonce?: number
  setNonce: Dispatch<SetStateAction<number | undefined>>

  safeTxGas?: number
  setSafeTxGas: Dispatch<SetStateAction<number | undefined>>
}>({
  setSafeTx: () => {},
  setSafeTxError: () => {},
  setNonce: () => {},
  setSafeTxGas: () => {},
})

// (err) => logError(Errors._103, (err as Error).message)

const SafeTxProvider = ({ children }: { children: ReactNode }): ReactElement => {
  const [safeTx, setSafeTx] = useState<SafeTransaction>()
  const [safeTxError, setSafeTxError] = useState<Error>()
  const [nonce, setNonce] = useState<number>()
  const [safeTxGas, setSafeTxGas] = useState<number>()

  // Signed txs cannot be updated
  const isSigned = safeTx && safeTx.signatures.size > 0

  // Recommended nonce and safeTxGas
  const recommendedParams = useRecommendedNonce(safeTx)

  // Priority to external nonce, then to the recommended one
  const finalNonce = nonce ?? recommendedParams?.nonce ?? safeTx?.data.nonce
  const finalSafeTxGas = safeTxGas ?? recommendedParams?.safeTxGas ?? safeTx?.data.safeTxGas

  // Update the tx when the nonce or safeTxGas change
  useEffect(() => {
    if (isSigned || !safeTx?.data) return
    if (safeTx.data.nonce === finalNonce && safeTx.data.safeTxGas === finalSafeTxGas) return

    createTx({ ...safeTx.data, safeTxGas: finalSafeTxGas }, finalNonce)
      .then(setSafeTx)
      .catch(setSafeTxError)
  }, [isSigned, finalNonce, finalSafeTxGas, safeTx?.data])

  return (
    <SafeTxContext.Provider
      value={{
        safeTx,
        safeTxError,
        setSafeTx,
        setSafeTxError,
        nonce,
        setNonce,
        safeTxGas,
        setSafeTxGas,
      }}
    >
      {children}
    </SafeTxContext.Provider>
  )
}

export default SafeTxProvider
