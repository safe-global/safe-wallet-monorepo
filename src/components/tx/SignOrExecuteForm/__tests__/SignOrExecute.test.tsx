import SignOrExecute from '../index'
import { render } from '@/tests/test-utils'
import * as hooks from '../hooks'
import type { TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
import type { SafeTxContextParams } from '@/components/tx-flow/SafeTxProvider'
import { SafeTxContext } from '@/components/tx-flow/SafeTxProvider'
import { createSafeTx } from '@/tests/builders/safeTx'

let isSafeOwner = true
// mock useIsSafeOwner
jest.mock('@/hooks/useIsSafeOwner', () => ({
  __esModule: true,
  default: jest.fn(() => isSafeOwner),
}))

// Mock proposeTx
jest.mock('@/services/tx/proposeTransaction', () => ({
  __esModule: true,
  default: jest.fn(() => Promise.resolve({ txId: '123' })),
}))

describe('SignOrExecute', () => {
  beforeEach(() => {
    isSafeOwner = true
    jest.clearAllMocks()
  })

  it('should display a loading component', () => {
    const { container } = render(<SignOrExecute onSubmit={jest.fn()} isExecutable={true} />)

    expect(container).toMatchSnapshot()
  })

  it('should display a confirmation screen', async () => {
    jest.spyOn(hooks, 'useProposeTx').mockReturnValue([
      {
        txInfo: {},
      } as TransactionDetails,
      undefined,
      false,
    ])

    const { container, getByTestId } = render(
      <SafeTxContext.Provider
        value={
          {
            safeTx: createSafeTx(),
          } as SafeTxContextParams
        }
      >
        <SignOrExecute onSubmit={jest.fn()} isExecutable={true} />
      </SafeTxContext.Provider>,
    )

    expect(getByTestId('sign-btn')).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })

  it('should display an error screen', async () => {
    jest.spyOn(hooks, 'useProposeTx').mockReturnValue([undefined, new Error('This is a mock error message'), false])

    const { container } = render(
      <SafeTxContext.Provider
        value={
          {
            safeTx: createSafeTx(),
          } as SafeTxContextParams
        }
      >
        <SignOrExecute onSubmit={jest.fn()} isExecutable={true} />
      </SafeTxContext.Provider>,
    )

    expect(container.querySelector('sign-btn')).not.toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})
