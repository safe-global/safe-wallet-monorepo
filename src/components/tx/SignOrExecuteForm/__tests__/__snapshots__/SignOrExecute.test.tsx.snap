// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignOrExecute should display a confirmation screen 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-118bpkk-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root cardContent css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1x8l8fd-MuiPaper-root-MuiAccordion-root"
          >
            <div
              aria-expanded="false"
              class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters accordion css-knbc2d-MuiButtonBase-root-MuiAccordionSummary-root"
              data-testid="decoded-tx-summary"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-6nefnh-MuiAccordionSummary-content"
              >
                Advanced details
                <span
                  class=""
                  data-mui-internal-clone-element="true"
                >
                  <mock-icon
                    aria-hidden="true"
                    classname="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-i8sqcg-MuiSvgIcon-root"
                    focusable="false"
                  />
                </span>
                <div
                  class="MuiBox-root css-i9gxme"
                />
              </div>
              <div
                class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17ceore-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                >
                  <div
                    class="MuiAccordion-region"
                    role="region"
                  >
                    <div
                      class="MuiAccordionDetails-root css-7wcoc8-MuiAccordionDetails-root"
                      data-testid="decoded-tx-details"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-118bpkk-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root cardContent css-46bh2p-MuiCardContent-root"
    >
      <div
        class="wrapper"
      >
        <div
          class="icon sign"
        >
          <mock-icon
            aria-hidden="true"
            classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-11dtfly-MuiSvgIcon-root"
            focusable="false"
          />
        </div>
        <div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-kmqua7-MuiTypography-root"
          >
            confirm
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body2 css-ceebrx-MuiTypography-root"
          >
            You're about to 
            create and 
            confirm
             this transaction.
          </p>
        </div>
      </div>
      <form>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth nestedDivider css-1cvjk81-MuiDivider-root"
        />
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
        >
          <div
            class="MuiStack-root css-irtfmw-MuiStack-root"
          >
            <span
              class=""
              data-mui-internal-clone-element="true"
            >
              <span
                data-track="batching: Add to batch"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-fj99we-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  <mock-icon
                    aria-hidden="true"
                    classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-gxmw0o-MuiSvgIcon-root"
                    focusable="false"
                  />
                  Add to batch
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </span>
            </span>
            <div
              class="MuiBox-root css-1aqmpzz"
            >
               
              <div
                class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-crkpmc-MuiDivider-root"
                role="separator"
              >
                <span
                  class="MuiDivider-wrapper css-165zycj-MuiDivider-wrapper"
                >
                  or
                </span>
              </div>
            </div>
            <span
              aria-label="Please connect your wallet"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1mhjei7-MuiButtonBase-root-MuiButton-root"
                data-testid="sign-btn"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                Sign
              </button>
            </span>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`SignOrExecute should display a loading component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-118bpkk-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root cardContent css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiBox-root css-17luq05"
      >
        <div
          class="box MuiBox-root css-0"
        >
          <div
            class="rect rectTl"
          />
          <div
            class="rect rectTr"
          />
          <div
            class="rect rectBl"
          />
          <div
            class="rect rectBr"
          />
          <div
            class="rect rectCenter"
          />
          <svg
            version="1.1"
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <filter
                id="gooey"
              >
                <fegaussianblur
                  in="SourceGraphic"
                  result="blur"
                  stdDeviation="3"
                />
                <fecolormatrix
                  in="blur"
                  mode="matrix"
                  result="goo"
                  values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 19 -9"
                />
                <fecomposite
                  in="SourceGraphic"
                  in2="goo"
                  operator="atop"
                />
              </filter>
            </defs>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SignOrExecute should display an error screen 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-118bpkk-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root cardContent css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiStack-root css-1sazv7p-MuiStack-root"
      >
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1x8l8fd-MuiPaper-root-MuiAccordion-root"
          >
            <div
              aria-expanded="false"
              class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters accordion css-knbc2d-MuiButtonBase-root-MuiAccordionSummary-root"
              data-testid="decoded-tx-summary"
              role="button"
              tabindex="0"
            >
              <div
                class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-6nefnh-MuiAccordionSummary-content"
              >
                Advanced details
                <span
                  class=""
                  data-mui-internal-clone-element="true"
                >
                  <mock-icon
                    aria-hidden="true"
                    classname="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-i8sqcg-MuiSvgIcon-root"
                    focusable="false"
                  />
                </span>
                <div
                  class="MuiBox-root css-i9gxme"
                />
              </div>
              <div
                class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17ceore-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
                >
                  <div
                    class="MuiAccordion-region"
                    role="region"
                  >
                    <div
                      class="MuiAccordionDetails-root css-7wcoc8-MuiAccordionDetails-root"
                      data-testid="decoded-tx-details"
                    >
                      <div
                        class="MuiGrid-root MuiGrid-container css-1d1otxt-MuiGrid-root"
                      >
                        <div
                          class="MuiGrid-root MuiGrid-item css-1wehkfy-MuiGrid-root"
                          data-testid="tx-row-title"
                          style="word-break: break-word;"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-fzss1y-MuiTypography-root"
                          >
                            safeTxGas:
                          </p>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                          data-testid="tx-data-row"
                        >
                          <div
                            class="MuiBox-root css-rurw5h"
                          >
                            0
                          </div>
                        </div>
                      </div>
                      <div
                        class="MuiGrid-root MuiGrid-container css-1d1otxt-MuiGrid-root"
                      >
                        <div
                          class="MuiGrid-root MuiGrid-item css-1wehkfy-MuiGrid-root"
                          data-testid="tx-row-title"
                          style="word-break: break-word;"
                        >
                          <p
                            class="MuiTypography-root MuiTypography-body1 css-fzss1y-MuiTypography-root"
                          >
                            Raw data:
                          </p>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                          data-testid="tx-data-row"
                        >
                          <div
                            class="MuiBox-root css-70qvj9"
                            data-testid="tx-data-row"
                          >
                            <div>
                              0
                               bytes
                            </div>
                            <span
                              aria-label="Copy to clipboard"
                              class=""
                              data-mui-internal-clone-element="true"
                              style="cursor: pointer;"
                            >
                              <button
                                aria-label="Copy to clipboard"
                                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-nmr6e8-MuiButtonBase-root-MuiIconButton-root"
                                tabindex="0"
                                type="button"
                              >
                                <mock-icon
                                  aria-hidden="true"
                                  classname="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-j1da6x-MuiSvgIcon-root"
                                  data-testid="copy-btn-icon"
                                  focusable="false"
                                />
                                <span
                                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                                />
                              </button>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiCard-root css-118bpkk-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root cardContent css-46bh2p-MuiCardContent-root"
    >
      <div
        class="wrapper"
      >
        <div
          class="icon sign"
        >
          <mock-icon
            aria-hidden="true"
            classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-11dtfly-MuiSvgIcon-root"
            focusable="false"
          />
        </div>
        <div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-kmqua7-MuiTypography-root"
          >
            confirm
          </h5>
          <p
            class="MuiTypography-root MuiTypography-body2 css-ceebrx-MuiTypography-root"
          >
            You're about to 
            create and 
            confirm
             this transaction.
          </p>
        </div>
      </div>
      <form>
        <hr
          class="MuiDivider-root MuiDivider-fullWidth nestedDivider css-1cvjk81-MuiDivider-root"
        />
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-dnrpxu-MuiCardActions-root"
        >
          <div
            class="MuiStack-root css-irtfmw-MuiStack-root"
          >
            <span
              class=""
              data-mui-internal-clone-element="true"
            >
              <span
                data-track="batching: Add to batch"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-fj99we-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  <mock-icon
                    aria-hidden="true"
                    classname="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-gxmw0o-MuiSvgIcon-root"
                    focusable="false"
                  />
                  Add to batch
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </span>
            </span>
            <div
              class="MuiBox-root css-1aqmpzz"
            >
               
              <div
                class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-crkpmc-MuiDivider-root"
                role="separator"
              >
                <span
                  class="MuiDivider-wrapper css-165zycj-MuiDivider-wrapper"
                >
                  or
                </span>
              </div>
            </div>
            <span
              aria-label="Please connect your wallet"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-1mhjei7-MuiButtonBase-root-MuiButton-root"
                data-testid="sign-btn"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                Sign
              </button>
            </span>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
